File: INA238.py
import smbus2 as smbus
import time as time

DEVICE_ID = 0x3F        # slave device ID register
CONFIG = 0x00           # Configuration (CONFIG) Register
ADC_CONFIG = 0x01       # ADC Configuration (ADC_CONFIG) Register
SHUNT_CAL = 0x02        # Shunt Voltage Measurement
VBUS = 0x05             # Bus Voltage Measurement
DIETEMP = 0x06          # Temperature Measurement (DIETEMP) Register
CURRENT = 0x07          # Current result
POWER = 0x08            # Power result

class Ina238:
    def __init__(self, addr, busNum):
        self.bus = smbus.SMBus(busNum) 
        self.addr = addr
        self.current_lsb = 0

        '''
        CONFIG reg: 1R00 0000 00R0 RRRR
                                 ^ADCRANGE selection
                      ^--------^ADC conversiton delay
                    ^system reset
        '''
        # do a system reset
        self.bus.write_word_data(addr, CONFIG, self.reverseWord(0x8000)) # system reset
        
        # set ADC conversion delay
        word = self.get16bitData(CONFIG)
        word &= (0xC07F << 6)   # clear bits 13-6
        word |= 0x40            # set ADC conversion delay to 2ms
        self.bus.write_word_data(addr, CONFIG, self.reverseWord(word))
        
        # set ADC averaging
        word = self.get16bitData(ADC_CONFIG)
        word &= (0xFFF8)    # clear bits 3-0
        word |= 0x03        # set ADC averaging to 64
        self.bus.write_word_data(addr, CONFIG, self.reverseWord(word))

    '''
    Reverses byte order to match INA238 transaction format
    '''
    def reverseWord(self, oldWord):
        newWord = oldWord & 0x00FF    # mask MSByte of oldWord    
        newWord = newWord << 8        # shift LSByte of newWord to MSByte   
        newWord |= (oldWord >> 8)     # shift MSByte of oldWord to LSByte and OR it to newWord
        return newWord

    def get16bitData(self, reg):
        '''
        reads 16-bit word from I2C bus using given slave adress 'addr',
        and register 'reg' to be read. Performs bitwise operations to arrange byte order
        returns integer value of the word.
        '''
        word = self.reverseWord(self.bus.read_word_data(self.addr, reg))
        return word
    
    def setShuntCal(self, res, maxCur):
        '''
        CURRENT_LSB = Max Expected Current / 2^15
                    = 0.050 A / 32768

        Use Equation 1 for ADCRANGE = 0 (±163.84 mV). or ADCRANGE = 1 (±40.96 mV), the value of SHUNT_CAL must be
        multiplied by 4

        Eq. 1- SHUNT_CAL = 819.2 x 10^6 x CURRENT_LSB x R_SHUNT (0.022R)
        '''
        self.current_lsb = maxCur / 32768
        factor = 1
        while(self.current_lsb < 1):
            self.current_lsb *= 10
            factor *= 10
        self.current_lsb = 10 / factor
        shunt_cal_value = round(819200000 * self.current_lsb * res)
        self.bus.write_word_data(self.addr, SHUNT_CAL, self.reverseWord(shunt_cal_value))

    # get Device ID
    def deviceID(self):
        word = self.get16bitData(DEVICE_ID)
        print("Device ID: 0x{:02x}" .format(word))

    # get Die Temprature
    def temprature(self):
        time.sleep(0.05)
        word = self.get16bitData(DIETEMP)
        print("Die Temp:", (word>>4)*0.125, "C") # shift 4 bits cause bits 0-3 are reserved

    # get Bus Voltage
    def voltage(self):
        time.sleep(0.05)
        word = self.get16bitData(VBUS)
        batVoltage = word*0.003125
        print("Bus Voltage: {:.{}f}".format(batVoltage, 2), 'V')
        return round(batVoltage, 2)

    # get Current
    def current(self):
        time.sleep(0.05)
        word = self.get16bitData(CURRENT)
        current = word*self.current_lsb*1000
        print("Current: {:.{}f}".format(current, 2), 'mA')
        return round(current, 2)

File: src/ina238.cpp
# include "ina238.hpp"

Ina238::Ina238(uint8_t addr, int busNum) {
    fd = wiringPiI2CSetup(addr);
    m_bus = busNum;
    m_addr = addr;
    m_currentLsb = 0;
    m_maxCurrent = 0;

    /*
    CONFIG reg: 1R00 0000 00R0 RRRR
                | |        | ^ADCRANGE selection
                | ^--------^ADC conversiton delay
                ^system reset
    */
    // do a system reset
    writeWordData(fd, CONFIG, 0x8000);
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    if(!checkDevice())
        return;

    /*
    ADC_CONFIG reg: 1111 1011 0110 1000
                    |  | | ||  ||  ^-^ADC averaging
                    |  | | ||  |^--^Temp conversion time
                    |  | | |^--^Shunt voltage conversion time
                    |  | ^-^Bus voltage conversion time
                    ^--^ADC mode
    */
    uint16_t word = getWordData(ADC_CONFIG);
    word &= 0xFFF8;     // clear bits 2-0
    word |= 0x04;      // set ADC averaging to 128
    writeWordData(fd, ADC_CONFIG, word);
    word = getWordData(ADC_CONFIG);
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
}

uint16_t Ina238::reverseWord(uint16_t oldWord) {
    uint16_t newWord = oldWord & 0x00FF;    // mask MSByte of oldWord    
    newWord = newWord << 8;                // shift LSByte of newWord to MSByte   
    newWord |= (oldWord >> 8);             // shift MSByte of oldWord to LSByte and OR it to newWord
    return newWord;
}

uint16_t Ina238::getWordData(uint8_t reg) {
    uint16_t word = static_cast<uint16_t>(reverseWord(wiringPiI2CReadReg16(fd, reg)));
    return word;
}

void Ina238::writeWordData(int fd, uint8_t reg, uint16_t data){
    wiringPiI2CWriteReg16(fd, reg, reverseWord(data));
}

void Ina238::setShuntCal(double res, double maxCur) {
    m_maxCurrent = maxCur;
    m_currentLsb = maxCur / 32768000.0f;
    double factor = 1.0f;
    while (m_currentLsb < 1.0f) {
        m_currentLsb *= 10.0f;
        factor *= 10.0f;
    }
    m_currentLsb = 1000.0f / factor;
    uint16_t shuntCalValue = static_cast<uint16_t>(std::round(819200000.0f * m_currentLsb * res));
    writeWordData(fd, SHUNT_CAL, shuntCalValue);
}

bool Ina238::checkDevice() {
    uint16_t word = getWordData(DEVICE_ID);
    if(word == 0x2381){
        std::cout << "Device is ready" << std::endl;
        return true;
    }
    else{
        std::cerr << "Device is not responding!" << std::endl;
        return false;
    }
}

void Ina238::deviceID() {
    uint16_t word = getWordData(DEVICE_ID);
    std::cout << "Device ID: 0x" << std::hex << word << std::endl;
}

void Ina238::temperature() {
    uint16_t word = getWordData(DIETEMP);
    float temp = static_cast<float>(word >> 4) * 0.125f; // shift 4 bits cause bits 0-3 are reserved
    std::cout << "Die Temp: " << temp << " C" << std::endl;
}

float Ina238::voltage() {
    uint16_t word = getWordData(ADC_CONFIG);
    ConversionTime_t conversionSelection = static_cast<ConversionTime_t>((word >> 9) & 0x07);
    float conversionTime = 0;
    unsigned int delay = 0;
    switch (conversionSelection){
        case t50us:
            conversionTime = 50; break;
        case t84us:
            conversionTime = 84; break;
        case t150us:
            conversionTime = 150; break;
        case t280us:
            conversionTime = 280; break;
        case t540us:
            conversionTime = 540; break;
        case t1052us:
            conversionTime = 1052; break;
        case t2074us:
            conversionTime = 2074; break;
    }
    short averaging = (word & 0x07);
    if(averaging <= 3)
        averaging = pow(2, averaging) * pow(2, averaging);
    else
        averaging = pow(2, averaging) * 8;
    
    delay = 7.13 * conversionTime * averaging; // no less than (7,13 * ADC conversion time * averaging)
    std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(delay/1000))); 
    
    word = getWordData(VBUS);
    float batVoltage = static_cast<float>(word) * 0.003125f;
    printf("Bus Voltage: %f V\n", batVoltage);
    return batVoltage;
}

float Ina238::current() {
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    uint16_t word = getWordData(CURRENT);
    double current = static_cast<double>(word) * m_currentLsb * 1000000.0f;
    if(current > m_maxCurrent)
        current = 0;
    std::cout << "Current: " << current << " uA" << std::endl;
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    return current;
}

uint16_t Ina238::getAlerts(){
    uint16_t word = getWordData(DIAG_ALRT);
    return word;
}